README
README for the man-db manual pager suite
========================================

Please read the man-db manual, included in the manual subdirectory of this
distribution.  It contains configuration details and other aspects of this
manual pager suite that are not duplicated or relevant in this README.
Check manual/README for details of the formatters required.

Read docs/INSTALL.autoconf for generic options to configure.
Read docs/INSTALL.quick if you know all about man-db.
Read NEWS for visible changes since the last public release.
Read ChangeLog for details of recent source code changes.
Read docs/TODO for future plans.

The C source requires an ANSI C compiler.


Copyright and licensing
=======================

Copyright (C) 1990, 1991 John W. Eaton.
Copyright (C) 1994, 1995 Markus Armbruster.
Copyright (C) 1994, 1995 Graeme W. Wilford. (Wilf.)
Copyright (C) 1995 Carl Edman.
Copyright (C) 1996, 1997, 1998, 2000 Fabrizio Polacco.
Copyright (C) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
              2011, 2012, 2013, 2014 Colin Watson.
Copyright (C) 1984, 1989, 1990, 1991, 1992, 1995, 1996, 1997, 1998, 1999,
              2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
              2010, 2011, 2012 Free Software Foundation, Inc.

man-db is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

man-db is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with man-db; if not, write to the Free Software Foundation,
Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

In addition, man-db incorporates Gnulib, copyrighted by the Free Software
Foundation and others. Note that much of Gnulib is distributed under the GNU
General Public License version 3 or later. This means that, although
man-db's own source code is licensed under GPL v2 or later, the work as a
whole falls under the terms of the GPL v3 or later. Unless you take special
pains to remove the GPL v3 portions, you must therefore follow the terms and
conditions of the GPL v3 or later when distributing man-db.


Notice regarding current state of FHS (Linux/?BSD)
==================================================

As of May 13th, 2001, the last public release of the Filesystem Hierarchy
Standard proposed the root of the manual page hierarchy as `/usr/share' and
the root of the writable cat hierarchy as `/var/cache/man' for the purposes
of man->cat filename translation.  As such, the following are defined in
./include/manconfig.h.in:

#define FHS_CAT_ROOT    "/var/cache/man"        /* required by fsstnd() */
#define FHS_MAN_ROOT    "/usr/share"            /* required by fsstnd() */

For compatibility with the old FSSTND, the following locations are also
defined:

#define CAT_ROOT        "/var/catman"           /* required by fsstnd() */
#define MAN_ROOT        "/usr"                  /* required by fsstnd() */

Should these locations change, simply define the paths accordingly and
recompile.  Other FHS changes relating to man/cat paths will not be
compatible with this version of man-db.

Non-generic arguments to configure
==================================

To allow the configuration program, configure, to be non-interactive, it can
be passed various options to alter the default settings.  Generic configure
options are discussed in docs/INSTALL.autoconf.  The following list of
options is extracted from the man-db manual.  It is strongly recommended
that relevant sections of the manual are read if any of these options are
used.

--enable-cache-owner[=ARG]
     By default, system-wide cache files will be owned by user man.  Use this
     option with an argument to change the cache file owner.

--disable-cache-owner
     Use this option to leave the ownership of system-wide cache files
     unconstrained.  Users will be allowed to modify them.

--disable-setuid
     By default, man will be installed as a setuid program to the user that
     owns the system-wide cache files.  Use this option to install man as a
     non-setuid program instead.

--enable-mandirs=OS
     By default, man-db supports manual page directories in any of several
     layouts used by free and proprietary versions of UNIX.  However, in
     certain cases, this can cause man-db to find the wrong page by mistake,
     especially when the names of some manual pages on the system contain
     periods.  Use this option with an argument of GNU, HPUX, IRIX, Solaris,
     or BSD (or more than one of these, separated by commas) to support only
     the layouts typically used on each of those systems.  Note that man-db is
     not currently capable of writing cat pages in the proper BSD layout.

--with-device=DEVICE
     Use this flag to alter the default output device used by NROFF. DEVICE is
     passed to NROFF with the -T option.  configure will test that NROFF will
     run with the supplied device argument.

--with-db=LIBRARY
     configure will look for database interface libraries in the order gdbm,
     Berkeley DB and finally ndbm and will #define appropriate variables
     relative to the first one found.  To override the built-in order on
     platforms having a choice of interface library, use this option to
     specify which library to use.

--enable-automatic-create
     If this flag is used, man will automatically create index databases for
     users' private manual page hierarchies.

--disable-automatic-update
     Normally, man will update entries in index databases if it finds newly
     installed manual pages (if the --update flag is used) or delete entries
     if manual pages are removed.  This flag suppresses this behaviour.

--disable-cats
     Normally, man will automatically try to create cat files corresponding to
     manual files when a manual page is read.  This flag suppresses this
     behaviour.


INSTALL
=======

Running configure.

	* READ `docs/INSTALL.autoconf' regarding ./configure options

	* RUN `./configure --help' to see what --enable and --with
	  options may be useful.

	* RUN `./configure' with the appropriate options and environment
	  variable settings

BROWSE or EDIT the following files that were created by the configuration
process.

	* `include/manconfig.h' regarding paths to support programs,
	  the default section list and other specific definitions.

	* `include/comp_src.h' if the default compressor support is
	  inadequate for your requirements.  (usually .Z [compress], 
	  .z, .gz [gzip])

configure will determine your system's ability to use native language
support (NLS) message catalogues.  You may set the environment variable
LINGUAS to limit the set of translations installed.  LINGUAS should contain
a space-separated list of two-letter language identifiers.  To compile
man-db with no support for message catalogues, simply pass the --disable-nls
option to configure.  N.B. This is not related to man's ability to display
NLS manual pages, support for which is compiled in by default. 

Running make.

	* RUN `make' to compile man-db with the set of translations chosen
	  when running `./configure'.

Sort out the man-db configuration file.

	* RUN `./src/man -l man/man5/manpath.5' from the root of this
	  distribution to read the man-db configuration file details.

	* EDIT `./src/man_db.conf' to your local requirements.

Install the package.

	* (gain superuser privileges for the rest of the steps)

	* RUN `make install' to install the utilities and manual pages.

Initialise the `index' databases for all manpaths marked as global in the
man-db configuration file.

	* RUN `mandb' (This step is equivalent to running straycats and 
	  makewhatis too).

The following steps are optional / dependent on local conventions.

	* ACKNOWLEDGE any warnings emitted by mandb. Bogus manual pages
	  are not included in the database and may be a waste of space.
	  Those pages without correctly formatted `whatis' lines are
	  included, but will have a whatis entry of "(unknown)"

	* CD tools and RUN `mkcatdirs -t' to see if you have all of the
	  required cat directories.  `mkcatdirs' without an option will 
	  display a usage message.

	* CD tools and RUN `checkman' with an argument of colon separated
	  manual page hierarchies to cross check for duplicated manual
	  pages.  If no argument is given, your default $MANPATH will be
	  used.
	
	  The output of checkman may be piped into a file and used as an
	  argument to `rm', the `is newer than' messages are directed to 
	  standard error.  e.g. `checkman > dups'

	  If you are confident that the duplicates found are indeed 
	  duplicates, you can back them up and delete them to save space.

	  At this point, running checkman again may yield further duplicates 
	  that were ignored the first time.
 
	* RUN `catman' with appropriate options to create any/all cat files
	  that you would like pre-formatted.


Multiple build directories
==========================

It is possible to build man-db in a directory other than the directory
containing this file (and all of the program sources).  This is particularly
useful if compiling on multiple architectures or testing various
configuration options as only a single copy of the sources is required.

To enable this support, simply change directory to where you would like to
build the package and run the configure program in this directory 
*from there*.  Further information about this support can be found in the 
generic install document `docs/INSTALL.autoconf'.


Makefile targets and variables
==============================

The standard GNU Makefile targets: all, install, uninstall, mostlyclean,
clean, distclean, realclean and TAGS are available in every Makefile-
supported directory.  In addition, the master Makefile has the dist target
to create a compressed and tarred distribution file.

During the configuration process, `configure' sets the installation
variables, `prefix' and `exec_prefix'.  These are then used to form other
variables such as `bindir' and `sysconfdir'.  To change any of these or 
other standard GNU install variables dynamically, issue the `make' command 
with variable expressions as arguments, eg. `make prefix=/usr/local/packages'

N.B. If prefix=/usr (either statically or dynamically), then sysconfdir=/etc
instead of the usual $(prefix)/etc.  To force sysconfdir to be /usr/etc, set
it on the make command line.


Default preprocessors
=====================

man-db uses a manual page directed preprocessor system, that is, each manual
page may request preprocessing by a selection of preprocessors.  Some
systems' manual pages do not come with this information built in.  In such
circumstances, it is advisable to set a default list of preprocessors that
each manual page should be passed through, so that those requiring special
processing are readable.  To achieve this, set DEFAULT_MANROFFSEQ (found in
include/manconfig.h) to the appropriate preprocessor string, after running
configure, but prior to compilation.  This is not necessary for the
following systems whose default preprocessing requirements are known.

  Known not to require DEFAULT_MANROFFSEQ:
    Linux, SunOS
  Known to require #define DEFAULT_MANROFFSEQ "t":
    Ultrix
  Known to require #define DEFAULT_MANROFFSEQ "te":
    HP-UX, OSF/1

If unsure of the default preprocessors required by a system, the standard
system's man(1) manual page may provide an answer. 


Contacting the maintainer
=========================

The current maintainer of man-db is Colin Watson <cjwatson@debian.org>.
Please feel free to contact me with any queries or problems you may have.
You can report bugs here:

  https://savannah.nongnu.org/bugs/?group=man-db


NEWS
man-db 2.9.5
============

Major changes since man-db 2.9.4:

	Move Git repository to GitLab (https://gitlab.com/cjwatson/man-db).

	Building man-db now requires a C99 compiler.

	Building man-db now requires Autoconf >= 2.64.

	Fixes:
	------

	* Manpath deduplication no longer mishandles the case where another
	  entry in the manpath is a suffix of a candidate path to append.

	* Fix potential crash in path searching if getcwd fails for reasons
	  other than running out of memory.

	* Fix crash in "globbing" test tool if run with no non-option
	  arguments.

man-db 2.9.4 (8 February 2021)
==============================

Major changes since man-db 2.9.3:

	Improvements:
	-------------

	* Recognise Romanian translations of the NAME section.

	* Treat "\[en]" (etc.) as another synonym for "\-" in NAME sections,
	  alongside the existing "\(en" (etc.).

	Fixes:
	------

	* Make the seccomp sandbox work better with libcs such as musl (S.
	  Gilles).

	* Make the seccomp sandbox allow clock_gettime64 as well as
	  clock_gettime (S. Gilles).

man-db 2.9.3 (22 June 2020)
===========================

Major changes since man-db 2.9.2:

	Fixes:
	------

	* Fix manual page translation infrastructure to compare po4a versions
	  with more than two components correctly.

	* Avoid incorrect markup in man(1) with po4a >= 0.58.

man-db 2.9.2 (1 June 2020)
==========================

Major changes since man-db 2.9.1:

	Fixes:
	------

	* Fix "man -X75-12" and "man -X100-12" to set the document font size
	  (using "-rS12") as well as the device (using "-TX75-12" or
	  "-TX100-12").

	* Fix incompatibility of "man -X" and friends with the seccomp
	  sandbox.

	Improvements:
	-------------

	* Add a bug tracker link to man-db's own manual pages.

	* Add support for zstd-compressed manual pages, thanks to Bernhard
	  Rosenkränzer.

man-db 2.9.1 (25 February 2020)
===============================

Major changes since man-db 2.9.0:

	Improvements:
	-------------

	* Drop fdutimens patch for GNU/Hurd; the bug it was working around
	  was fixed in glibc 2.28.

	* Add MANDB_MAP entry mapping /snap/man to /var/cache/man/snap.

man-db 2.9.0 (23 October 2019)
==============================

Major changes since man-db 2.8.7:

	Fixes:
	------

	* man --recode and manconv now adjust encoding declarations on the
	  first line of their input to refer to the new encoding.

	* Fix comparison of candidate manual pages to correctly handle the case
	  where the language elements are the same and match the locale, but
	  the territory elements differ.

	Improvements:
	-------------

	* Many typographical improvements to man-db's own manual pages,
	  largely thanks to Bjarni Ingi Gislason.

	* Rewrite parts of man(1) to make it a more accessible introduction.

	* If run with no arguments or only a section, man now suggests
	  running "man man".

	* man now understands the <page>(<section>) form on its command
	  line, so for example "man 'chmod(2)'" is now the same as "man 2
	  chmod".  While this requires more quoting, it may also be more
	  convenient when copying and pasting cross-references to manual
	  pages.

	* manconv now guesses the input encoding based on the file name if
	  it is not explicitly specified.

	* There is a new "man-recode" program.  It fulfils a purpose similar
	  to "man --recode", but has an interface designed for bulk
	  conversion and so can be much faster when used on a large number
	  of pages.

	Compatibility notes:
	--------------------

	* Remove the ability to undefine COMP_SRC at build-time to disable
	  reading compressed manual pages.  This was always only a
	  micro-optimisation, and it wasn't worth the extra code complexity.
	  You can still configure without any compressors (either by not
	  having them available at build time, or by configuring with
	  "--with-gzip=" etc.) to achieve much the same practical effect,
	  although the test suite still requires at least gunzip.

man-db 2.8.7 (26 August 2019)
=============================

Major changes since man-db 2.8.6.1:

	Fixes:
	------

	* Further workarounds for ESET File Security: allow sendmsg when it
	  is in use.

	* The seccomp sandbox now causes disallowed system calls to return
	  EPERM rather than raising SIGSYS, in the hope of being less
	  disruptive to preload hacks.

	* Make seccomp sandbox allow getrandom, used by Hardened Malloc.

	* man no longer saves cat pages if --no-hyphenation or
	  --no-justification is used.

	* Move decompression code out of libman.  This should really fix a
	  link failure using the Darwin linker (unsuccessfully attempted in
	  2.8.6), and possibly on other platforms too.

	* Pass the database file name around in function parameters, rather
	  than storing it in a global variable with an unresolved symbol
	  from libmandb.  Like the previous item, this fixes a link failure
	  using the Darwin linker, and possibly on other platforms too.

	* Return database entries in sorted order when using NDBM.  (This is
	  based on a similar fix to the GDBM backend in man-db 2.4.2.)

	Improvements:
	-------------

	* Recognise Esperanto, Tamil, and Ukrainian translations of the NAME
	  section.

man-db 2.8.6.1 (5 August 2019)
==============================

Major changes since man-db 2.8.6:

	Fixes:
	------

	* Fix missing memory copies in ult_src that caused segfaults in
	  mandb.

man-db 2.8.6 (3 August 2019)
============================

Major changes since man-db 2.8.5:

	Fixes:
	------

	* If more than one of "../man", "man", "../share/man", and
	  "share/man" exist relative to a directory on $PATH, then all of
	  them are now added to the automatically-determined manpath;
	  previously, only the first was considered.

	* Remove arbitrary limit on manpath size.

	* The systemd database maintenance service now runs mandb with the
	  --quiet option, avoiding excess log messages.

	* Default to --without-systemdsystemunitdir and
	  --without-systemdtmpfilesdir on non-Linux systems.


lib/util.c
/*
 * util.c
 *
 * Copyright (C) 1990, 1991 John W. Eaton.
 * Copyright (C) 1994, 1995 Graeme W. Wilford. (Wilf.)
 * Copyright (C) 2001, 2002, 2004, 2007, 2008, 2010 Colin Watson.
 *
 * This file is part of man-db.
 *
 * man-db is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * man-db is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with man-db; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * John W. Eaton
 * jwe@che.utexas.edu
 * Department of Chemical Engineering
 * The University of Texas at Austin
 * Austin, Texas  78712
 *
 * Wed May  4 15:44:47 BST 1994 Wilf. (G.Wilford@ee.surrey.ac.uk): slight
 * changes to all routines, mainly cosmetic.
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif /* HAVE_CONFIG_H */

#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <unistd.h>
#include <locale.h>

#include "attribute.h"
#include "stat-time.h"
#include "timespec.h"
#include "xvasprintf.h"

#include "gettext.h"

#include "manconfig.h"

#include "error.h"
#include "pipeline.h"

/*
 * Does file a have a different timestamp to file b?
 *
 * case:
 *
 *   a is man_page, b is cat_page
 *
 *   a and b have different times  returns  1/3  (ret & 1) == 1
 *   a and b have same times       returns  0/2  (!(ret & 1)) == 1
 *   a is zero in length           returns  + 2 (for Wilf. and his stray cats)
 *   b is zero in length           returns  + 4
 *   stat on a fails               returns   -1
 *   stat on b fails               returns   -2
 *   stat on a and b fails         returns   -3
 */
int is_changed (const char *fa, const char *fb)
{
	struct stat fa_sb;
	struct stat fb_sb;
	int fa_stat;
	int fb_stat;
	int status = 0;

	debug ("is_changed: a=%s, b=%s", fa, fb);

	fa_stat = stat (fa, &fa_sb);
	if (fa_stat != 0)
		status = 1;

	fb_stat = stat (fb, &fb_sb);
	if (fb_stat != 0)
		status |= 2;

	if (status != 0) {
		debug (" (%d)\n", -status);
		return -status;
	}

	if (fa_sb.st_size == 0)
		status |= 2;

	if (fb_sb.st_size == 0)
		status |= 4;

	status |= (timespec_cmp (get_stat_mtime (&fa_sb),
				 get_stat_mtime (&fb_sb)) != 0);

	debug (" (%d)\n", status);
	return status;
}

/*
 * Is path a directory?
 */
int is_directory (const char *path)
{
	struct stat sb;
	int status;

	status = stat (path, &sb);

	if (status != 0)
		return status;

	return ((sb.st_mode & S_IFDIR) != 0);
}

/* Escape dangerous metacharacters before dumping into a shell command. */
char *escape_shell (const char *unesc)
{
	char *esc, *escp;
	const char *unescp;

	if (!unesc)
		return NULL;

	escp = esc = xmalloc (strlen (unesc) * 2 + 1);
	for (unescp = unesc; *unescp; unescp++)
		if ((*unescp >= '0' && *unescp <= '9') ||
		    (*unescp >= 'A' && *unescp <= 'Z') ||
		    (*unescp >= 'a' && *unescp <= 'z') ||
		    strchr (",-./:@_", *unescp))
			*escp++ = *unescp;
		else {
			*escp++ = '\\';
			*escp++ = *unescp;
		}
	*escp = 0;
	return esc;
}

/* Remove a directory and all files in it.  Only recurse beyond that if
 * RECURSE is set.
 */
int remove_directory (const char *directory, int recurse)
{
	DIR *handle = opendir (directory);
	struct dirent *entry;

	if (!handle)
		return -1;
	while ((entry = readdir (handle)) != NULL) {
		struct stat st;
		char *path;

		if (STREQ (entry->d_name, ".") || STREQ (entry->d_name, ".."))
			continue;
		path = xasprintf ("%s/%s", directory, entry->d_name);
		assert (path);
		if (stat (path, &st) == -1) {
			free (path);
			closedir (handle);
			return -1;
		}
		if (recurse && S_ISDIR (st.st_mode)) {
			if (remove_directory (path, recurse) == -1) {
				free (path);
				closedir (handle);
				return -1;
			}
		} else if (S_ISREG (st.st_mode)) {
			if (unlink (path) == -1) {
				free (path);
				closedir (handle);
				return -1;
			}
		}
		free (path);
	}
	closedir (handle);

	if (rmdir (directory) == -1)
		return -1;
	return 0;
}

/* Returns an allocated copy of s, with leading and trailing spaces
 * removed.
 */
char * ATTRIBUTE_MALLOC trim_spaces (const char *s)
{
	int length;
	while (*s == ' ')
		++s;
	length = strlen (s);
	while (length && s[length - 1] == ' ')
		--length;
	return xstrndup (s, length);
}

char *lang_dir (const char *filename)
{
	char *ld;	/* the lang dir: point to static data */
	const char *fm;	/* the first "/man/" dir */
	const char *sm;	/* the second "/man?/" dir */

	ld = xstrdup ("");
	if (!filename)
		return ld;

	/* Check whether filename is in a man page hierarchy. */
	if (STRNEQ (filename, "man/", 4))
		fm = filename;
	else {
		fm = strstr (filename, "/man/");
		if (fm)
			++fm;
	}
	if (!fm)
		return ld;
	sm = strstr (fm + 2, "/man");
	if (!sm)
		return ld;
	if (sm[5] != '/')
		return ld;
	if (!strchr ("123456789lno", sm[4]))
		return ld;

	/* If there's no lang dir element, it's an English man page. */
	if (sm == fm + 3) {
		free (ld);
		return xstrdup ("C");
	}

	/* found a lang dir */
	fm += 4;
	sm = strchr (fm, '/');
	if (!sm)
		return ld;
	free (ld);
	ld = xstrndup (fm, sm - fm);
	debug ("found lang dir element %s\n", ld);
	return ld;
}

void init_locale (void)
{
	char *locale = setlocale (LC_ALL, "");
	if (!locale &&
	    !getenv ("MAN_NO_LOCALE_WARNING") &&
	    !getenv ("DPKG_RUNNING_VERSION"))
		/* Obviously can't translate this. */
		error (0, 0, "can't set the locale; make sure $LC_* and $LANG "
			     "are correct");
	setenv ("MAN_NO_LOCALE_WARNING", "1", 1);
#ifdef ENABLE_NLS
	bindtextdomain (PACKAGE, LOCALEDIR);
	bindtextdomain (PACKAGE "-gnulib", LOCALEDIR);
	textdomain (PACKAGE);
#endif /* ENABLE_NLS */
}
